### PR TITLE
Anej/cell data

### PR DESCRIPTION
**Pull request by AnejSvete**
This PR contains a few modifications to the codebase to make it more applicable to general datasets which are not implemented directly in `latte`.
Furthermore, it introduces the change where, instead of investigating complements of subspaces in the full representation space, `latte` now computes an orthogonal basis of the complement to reduce the dimensionality.
It also fixes a few bugs.

### CODE REVIEW

File: latte/dataset/dsprites.py [hunk 1]
```diff
@@ -52,6 +52,10 @@ class DSpritesFactor(IntEnum):
 }
 
 
+attribute_names = ["shape", "scale", "orientation", "position_x", "position_y"]
```
Comment (pawel-czyz) [inline latte/dataset/dsprites.py:55]: Nice :+1:

File: latte/models/cca/estimation.py [hunk 2]
```diff
@@ -56,7 +56,9 @@ def fit(X: np.ndarray, Z: np.ndarray, subspace_size: int, p_train: float = 0.8)
     cca.fit(X_train, Z_train)
 
     A_hat_1, A_hat_2 = cca.x_weights_, cca.y_weights_
-    E_hat_1, E_hat_2 = np.eye(A_hat_1.shape[0]) - A_hat_1 @ A_hat_1.T, np.eye(A_hat_2.shape[0]) - A_hat_2 @ A_hat_2.T
+    (d_1, m_1), (d_2, m_2) = A_hat_1.shape, A_hat_2.shape
+    E_hat_1 = np.linalg.svd(np.eye(A_hat_1.shape[0]) - A_hat_1 @ A_hat_1.T, full_matrices=False)[0][:, : d_1 - m_1]
```
Comment (pawel-czyz) [inline latte/models/cca/estimation.py:60]: Minor thing: why not `np.eye(d_1)`?

File: latte/models/cca/estimation.py [hunk 3]
```diff
@@ -56,7 +56,9 @@ def fit(X: np.ndarray, Z: np.ndarray, subspace_size: int, p_train: float = 0.8)
...
+    E_hat_2 = np.linalg.svd(np.eye(A_hat_2.shape[0]) - A_hat_2 @ A_hat_2.T, full_matrices=False)[0][:, : d_2 - m_2]
```
Comment (pawel-czyz) [inline latte/models/cca/estimation.py:61]: (As above, `np.eye(d_2)` may read better).

File: latte/models/mist/estimation.py [hunk 4]
```diff
@@ -381,8 +383,8 @@ def fit(
         estimator=best_model,
         A_1=A_hat_x,
         A_2=A_hat_z,
-        E_1=torch.eye(A_hat_x.shape[0]) - A_hat_x @ A_hat_x.T,
-        E_2=torch.eye(A_hat_z.shape[0]) - A_hat_z @ A_hat_z.T,
+        E_1=torch.linalg.svd(torch.eye(A_hat_x.shape[0]) - A_hat_x @ A_hat_x.T, full_matrices=False)[0][:, : d_x - m_x],
```
Comment (pawel-czyz) [inline latte/models/mist/estimation.py:386]: Nice, again I'd use `d_x` and `d_z`.

File: latte/models/probabilistic_pca/probabilistic_pca.py [hunk 5]
```diff
@@ -64,6 +71,8 @@ def generate(
     Returns:
         Populated GaussianGenerativeDataset
     """
+    if not d_measured:
```
Comment (pawel-czyz) [inline latte/models/probabilistic_pca/probabilistic_pca.py:74]: I don't think it'll be troublesome, but in principle checking for None is accomplished by:
`if d_measured is None:`

However, one can use `d_measured=0` and this `if` will be executed (which makes sense). I'd suggest adding an `assert` if you decided that you didn't like this behaviour (or if you want to get rid of negative values)

File: latte/models/rlace/estimation.py [hunk 6]
```diff
@@ -60,13 +65,25 @@ def fit(X: np.ndarray, Z: np.ndarray, rlace_params: Dict[str, Any], p_train: flo
     X_test_projected_A, X_test_projected_E = X_test @ A_hat_x, X_test @ E_hat_x
 
     # Construct the result
+    rng = np.random.default_rng(rng)
+    mi_estimation_ix = rng.choice(len(Z_test), min(8000, len(Z_test)), replace=False)
```
Comment (pawel-czyz) [inline latte/models/rlace/estimation.py:69]: This hard-coded `8000` looks a bit fishy. Can it be made a function parameter with a default value?

File: latte/models/cca/estimation.py (hunk 3 above)
Comment (pawel-czyz) [reply latte/models/cca/estimation.py:61]: Actually, I've noticed that the pattern
`svd(eye(x) - y)[0][:, :, z]`
appears several times – perhaps making it into a subroutine (with a comment on `[0][:, :, z]`) would be helpful :slightly_smiling_face:

File: latte/models/rlace/estimation.py [hunk 7]
```diff
@@ -60,13 +65,25 @@ def fit(X: np.ndarray, Z: np.ndarray, rlace_params: Dict[str, Any], p_train: flo
...
-        mutual_information=naive_ksg.mutual_information([X_test_projected_A.numpy(), Z_test.reshape(-1, 1)], k=3),
-        complement_mutual_information=naive_ksg.mutual_information(
-            [X_test_projected_E.numpy(), Z_test.reshape(-1, 1)], k=3
-        ),
+        mutual_information=ksg.estimate_mi_ksg(
+            X_test_projected_A[mi_estimation_ix].numpy(),
+            Z_test[mi_estimation_ix].reshape(-1, 1).numpy()
+            if isinstance(Z_test, torch.Tensor)
+            else Z_test[mi_estimation_ix].reshape(-1, 1),
+            neighborhoods=(5,),
```
Comment (pawel-czyz) [inline latte/models/rlace/estimation.py:76]: I'd refrain from hard-coded `5`, especially that now we have [an implementation of nice interface](https://github.com/pawel-czyz/bmi/blob/main/src/bmi/estimators/ksg.py). Could you please add a TODO comment with my name so I don't forget to change this?

File: scripts/find_vae_image_representation_subspaces.py [hunk 8]
```diff
@@ -257,7 +258,7 @@ def _process_subspace(
     rng = np.random.default_rng(rng)
 
     evaluation_ixs = rng.choice(len(Z), size=cfg.estimation_N, replace=False)
-    latte_ksg_evaluation_ixs = rng.choice(len(Z), size=2000, replace=False)
+    latte_ksg_evaluation_ixs = rng.choice(len(Z), size=4000, replace=False)
```
Comment (pawel-czyz) [inline scripts/find_vae_image_representation_subspaces.py:261]: Question: why is here 4000? In the other file, the default is 8,000, isn't it?

**Review by pawel-czyz — APPROVED** [review]
Had some minor  omments, but overall it's very nice (as always)! :slightly_smiling_face:

File: latte/models/cca/estimation.py (hunk 2 above)
Comment (AnejSvete) [reply latte/models/cca/estimation.py:60]: Good catch, will change! :)

File: latte/models/rlace/estimation.py (hunk 6 above)
Comment (AnejSvete) [reply latte/models/rlace/estimation.py:69]: Will do :)

File: scripts/find_vae_image_representation_subspaces.py (hunk 8 above)
Comment (AnejSvete) [reply scripts/find_vae_image_representation_subspaces.py:261]: Yes, that was a bit of a relic from testing. I will make it a parameter as above.

File: latte/models/rlace/estimation.py (hunk 7 above)
Comment (AnejSvete) [reply latte/models/rlace/estimation.py:76]: Will do!

**Comment by pawel-czyz** [comment]
Wonderful! Feel free to merge :slightly_smiling_face:

**Pull request merged**